server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/factory/PromiseMap.java

```
@@ -0,0 +1,116 @@
+package org.jboss.as.clustering.infinispan.cs.factory;
+
+import org.jboss.as.clustering.infinispan.InfinispanLogger;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A map implementation based on promises. {@link #get(Object, long, java.util.concurrent.TimeUnit)} method will
+ * block until a corresponding {@link #put(Object, Object)} is called or a timeout occurs.
+ *
+ * @author Sebastian Laskawiec
+ * @since 8.0
+ */
+public class PromiseMap<K, V> implements Map<K, V> {
+
+   private static final TimeUnit DEFAULT_WAIT_TIME_UNIT = TimeUnit.MILLISECONDS;
+   private static final long DEFAULT_WAIT_TIMEOUT = 100;
+
+   private ConcurrentHashMap<K,CompletableFuture<V>> internalMap = new ConcurrentHashMap<>();
+
+   @Override
+   public int size() {
+      return internalMap.size();
+   }
+
+   @Override
+   public boolean isEmpty() {
+      return internalMap.isEmpty();
+   }
+
+   @Override
+   public boolean containsKey(Object key) {
+      return internalMap.containsKey(key);
+   }
+
+   @Override
+   public boolean containsValue(Object value) {
+      return true;
```

This should throw an exception